File: 프로그래머스/2/176962. 과제 진행하기/과제 진행하기.py
def solution(plans):
    answer = []
    
    for i in range(len(plans)):
        s= plans[i][1].split(':')
        plans[i][1] = int(s[0])*60+int(s[1])
        plans[i][2] = int(plans[i][2])
    
    plans.sort(key= lambda x:x[1])
    stack = []
    stack.append(plans[0])
    
    now = plans[0][1]
    for i in range(1, len(plans)):
        lang, start, spend = plans[i]
        while stack:
            sl, ss, sp = stack.pop()

            if now <= ss:
                now = ss
            
            finish = now + sp
            
            if finish <= start:
                answer.append(sl)
                now += sp
            else:
                stack.append([sl, ss, finish - start])
                now = start
                break
        stack.append(plans[i])

    while stack:
        l = stack.pop()[0]
        answer.append(l)

    return answer

File: 프로그래머스/2/176962. 과제 진행하기/과제 진행하기.java
import java.util.*;
class Solution {
    public String[] solution(String[][] plans) {
        String[] answer = {};
        
        Arrays.sort(plans, (a,b) -> time(a[1]) - time(b[1]));
        
        for(String[] p: plans){
            System.out.println(p[1]);
        }
        Queue<String[]> q = new LinkedList<>();
        int n = plans.length;
        
        for(int i =0; i<n; i++){
            plans[i][1] = String.valueOf(time(plans[i][1]));
            q.offer(plans[i]);
        }
        
        Stack<String[]> stack = new Stack<>();
        int nowTime = Integer.parseInt(q.peek()[1])+Integer.parseInt(q.peek()[2]);
        List<String> res = new ArrayList<>();
        while(!q.isEmpty()){
            String[] now = q.poll();
            String name = now[0];
            int st = Integer.parseInt(now[1]);
            int d = Integer.parseInt(now[2]);
            
            if(!q.isEmpty()){
                int nextTime = Integer.parseInt(q.peek()[1]);
                
                if(st+d <= nextTime){
                    res.add(name);
                    nowTime = st+d;
                    
                    while(!stack.isEmpty()){
                        String[] p = stack.pop();
                        String pname = p[0];
                        int ptime = Integer.parseInt(p[1]);
                        
                        if(nowTime+ptime <= nextTime){
                            res.add(pname);
                            nowTime += ptime;
                        }else{
                            p[1] = String.valueOf(ptime-(nextTime - nowTime));
                            stack.push(p);
                            break;
                        }
                    }
                }else{
                    String[] p = new String[]{name, String.valueOf(d- (nextTime-st))};
                    stack.push(p);
                }
            }else{
                res.add(name);
                while (!stack.isEmpty()) {
                    res.add(stack.pop()[0]);
                }
            }
            
        }
        
        
        return res.toArray(new String[0]);
    }
    private static int time(String t){
        String[] tt = t.split(":");
        return Integer.parseInt(tt[0])*60 + Integer.parseInt(tt[1]);
    }
}